pkg/controller/controller_test.go

```diff
@@ -1014,6 +1017,185 @@ var _ = ginkgo.Describe("CSIControllerService", func() {
+			req.Parameters[common.KeyNasName] = "nasA"
+
+			res, err := ctrlSvc.CreateVolume(context.Background(), req)
+			gomega.Expect(err).To(gomega.BeNil())
+			gomega.Expect(res).To(gomega.Equal(&csi.CreateVolumeResponse{
+				Volume: &csi.Volume{
+					CapacityBytes: validVolSize,
+					VolumeId:      filepath.Join(validBaseVolID, secondValidID, "nfs"),
+					VolumeContext: map[string]string{
+						common.KeyArrayVolumeName:   "my-vol",
+						common.KeyProtocol:          "nfs",
+						common.KeyArrayID:           secondValidID,
+						common.KeyNfsACL:            "A::OWNER@:RWX",
+						common.KeyNasName:           "nasA",
+						common.KeyVolumeDescription: req.Name + "-" + validNamespaceName,
+						common.KeyServiceTag:        validServiceTag,
+						KeyCSIPVCName:               req.Name,
+						KeyCSIPVCNamespace:          validNamespaceName,
+					},
+					AccessibleTopology: []*csi.Topology{{Segments: map[string]string{common.Name + "/" + ctrlSvc.Arrays()[secondValidID].GetIP() + "-nfs": "true"}}},
+				},
+			}))
+		})
+
+		ginkgo.It("should fail when there is no least used NAS[Invalid NAS]", func() {
+			clientMock.On("GetNASServers", mock.Anything).Return([]gopowerstore.NAS{invalidNAS4}, nil)
+
+			req := getTypicalCreateVolumeNFSRequest("my-vol", validVolSize)
+			req.Parameters[common.KeyArrayID] = secondValidID
+
+			req.Parameters[KeyCSIPVCName] = req.Name
+			req.Parameters[KeyCSIPVCNamespace] = validNamespaceName
+			req.Parameters[common.KeyNasName] = "nasA, nasB, nasC, nasX"
+
+			res, err := ctrlSvc.CreateVolume(context.Background(), req)
+			gomega.Expect(err.Error()).To(gomega.ContainSubstring("no suitable NAS server found, please ensure the NAS is running and healthy"))
+			gomega.Expect(res).To(gomega.BeNil())
+		})
+
+		ginkgo.It("should fail when fs creation limit is exceeded", func() {
+			clientMock.On("GetNASServers", mock.Anything).Return([]gopowerstore.NAS{validNAS1, validNAS2, validNAS3, invalidNAS4}, nil)
+			clientMock.On("GetNASByName", mock.Anything, "nasA").Return(gopowerstore.NAS{ID: validNasID}, nil)
+			clientMock.On("CreateFS", mock.Anything, mock.Anything).
+				Return(gopowerstore.CreateResponse{}, gopowerstore.APIError{
+					ErrorMsg: &api.ErrorMsg{
+						StatusCode: http.StatusUnprocessableEntity,
+						Message:    "New file system can not be created. The limit of 125 file systems for the NAS server 24aefac2-a796-47dc-886a-c73ff8c1a671 has been reached.",
+					},
+				})
+
+			req := getTypicalCreateVolumeNFSRequest("my-vol", validVolSize)
+			req.Parameters[common.KeyArrayID] = secondValidID
+
+			req.Parameters[KeyCSIPVCName] = req.Name
+			req.Parameters[KeyCSIPVCNamespace] = validNamespaceName
+			req.Parameters[common.KeyNasName] = "nasA, nasB, nasC, nasX"
+
+			res, err := ctrlSvc.CreateVolume(context.Background(), req)
+			gomega.Expect(err.Error()).To(gomega.ContainSubstring("The limit of 125 file systems for the NAS server"))
+			gomega.Expect(res).To(gomega.BeNil())
```

Comment:
It would be great to double-check both the nas failure count and cooldown to make sure everything is working as expected.